avi/resource_avi_cloud.go

```diff
@@ -236,10 +243,104 @@ func ResourceAviCloudRead(d *schema.ResourceData, meta interface{}) error {
+	var robj interface{}
+	var err error
+	path := "api/vimgrvcenterruntime?cloud_uuid=" + cloudUuid
+	i := 0
+	for ; i < maxRetry; i++ {
+		if err = client.AviSession.Get(path, &robj); err == nil {
+			objCount := robj.(map[string]interface{})["count"].(float64)
+			if objCount == float64(1) {
+				inventoryState := robj.(map[string]interface{})["results"].([]interface{})[0].(map[string]interface{})["inventory_state"].(string)
+				if isValidCloudState(inventoryState) {
+					log.Printf("Got expected inventory state %s", inventoryState)
+					break
+				} else {
+					log.Printf("Didn't get expected inventory state. Current state is %s", inventoryState)
+				}
+			} else {
+				log.Printf("Inventory is not completed")
+			}
+		} else {
+			log.Printf("[Error] Got error while retrieving vimgrvcenterruntime %s", err.Error())
+		}
+		time.Sleep(10 * time.Second)
+	}
+	if i == maxRetry && err == nil {
+		err = errors.New("didn't get expected vcenter(vimgrvcenterruntime) inventory state")
+	}
+	return err
+}
+
+// Wait untill cloud state CLOUD_STATE_PLACEMENT_READY
+func checkCloudState(cloudUuid string, client *clients.AviClient, maxRetry int) error {
+	var robj interface{}
+	var err error
+	path := "api/cloud-inventory?uuid=" + cloudUuid
+	i := 0
+	for ; i < maxRetry; i++ {
+		if err = client.AviSession.Get(path, &robj); err == nil {
+			objCount := robj.(map[string]interface{})["count"].(float64)
+			if objCount == float64(1) {
+				cloudState := robj.(map[string]interface{})["results"].([]interface{})[0].(map[string]interface{})["status"].(map[string]interface{})["state"].(string)
+				if cloudState == "CLOUD_STATE_PLACEMENT_READY" {
+					break
+				} else {
+					log.Printf("Didn't get expected cloud state. Current cloud state is %s", cloudState)
+				}
+			} else {
+				log.Printf("Didn't get inventory for cloud")
+			}
+		} else {
+			log.Printf("[Error] Got Error while retrieving cloud-inventory %s", err.Error())
+		}
+		time.Sleep(10 * time.Second)
+	}
+	if i == maxRetry && err == nil {
+		err = errors.New("didn't get expected state CLOUD_STATE_PLACEMENT_READY in cloud-inventory")
+	}
+	return err
+}
+
 func resourceAviCloudCreate(d *schema.ResourceData, meta interface{}) error {
 	s := ResourceCloudSchema()
-	err := ApiCreateOrUpdate(d, meta, "cloud", s)
-	if err == nil {
+	var err error
+	cloudType := d.Get("vtype")
+	vcenterConfig, isVcenterConfig := d.GetOk("vcenter_configuration")
+	if cloudType == "CLOUD_VCENTER" && isVcenterConfig {
+		maxRetry := d.Get("cloud_state_max_retry").(int)
+		mgmtNetwork := vcenterConfig.(*schema.Set).List()[0].(map[string]interface{})["management_network"].(string)
+		mgmtNetwork = "vimgrruntime?name=" + mgmtNetwork
+		if err = ApiCreateOrUpdate(d, meta, "cloud", s); err == nil {
+			if err = ResourceAviCloudRead(d, meta); err == nil {
+				client := meta.(*clients.AviClient)
+				uuid := d.Get("uuid").(string)
+				if err = waitForVcenterState(uuid, client, maxRetry); err == nil {
```

Comment:
I suggest we refactor this code to move to another method where there are not much nesting. 
In that function it can wait, re-apply the configuration.

Also, can we do what we do in ansible where we can pop the network configuration setting in the intial creation and then set that again?